File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        index = {}
        l = 0
        res = 0

        for r in range(len(s)):
            if s[r] in index:
                l = max(l, index[s[r]] + 1)
            index[s[r]] = r
            res = max(res, r - l + 1)
        return res

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.swift
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     var val: Int
 *     var left: TreeNode?
 *     var right: TreeNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func goodNodes(_ root: TreeNode?) -> Int {

        func dfs(_ root: TreeNode?, _ maxVal: Int) -> Int {
            guard let root = root else { return 0 }

            var res = 0
            if root.val >= maxVal {
                res = 1
            }
            let newMax = max(maxVal, root.val)
            res += dfs(root.left, newMax)
            res += dfs(root.right, newMax)
            return res
        }
        return dfs(root, root?.val ?? Int.min)
    }
}
